src/neuron/cache/mechanism_range.hpp

```diff
@@ -0,0 +1,154 @@
+#pragma once
+#include "membfunc.h"
+#include "nrn_ansi.h"
+#include "nrnoc_ml.h"
+
+#include <array>
+
+namespace neuron::cache {
+/**
+ * @brief Call the given method with each dparam index that should be cached for a mechanism.
+ *
+ * The callable will be invoked with the largest index first. This is useful if you want to resize a
+ * container and use the indices as offsets into it.
+ */
+template <typename Callable>
+void indices_to_cache(short type, Callable callable) {
+    auto const pdata_size = nrn_prop_dparam_size_[type];
+    auto* const dparam_semantics = memb_func[type].dparam_semantics;
+    for (int field = pdata_size - 1; field >= 0; --field) {
+        // Check if the field-th dparam of this mechanism type is an ion variable. See
+        // hoc_register_dparam_semantics.
+        auto const sem = dparam_semantics[field];
+        // TODO is area constant? could we cache the value instead of a pointer to it?
+        // TODO ion type can be handled differently
+        if ((sem > 0 && sem < 1000) || sem == -1 /* area */) {
+            std::invoke(callable, field);
+        }
+    }
+}
+
+/**
+ * @brief Version of Memb_list for use in performance-critical code.
+ *
+ * Unlike Memb_list, this requires that the number of fields is known at compile time.
+ * This is typically only true in translated MOD file code. The idea is that an instance of this
+ * class will be created outside a loop over the data vectors and then used inside the loop.
+ *
+ * It is the responsibility of the caller to ensure that the model remains sorted beyond the
+ * lifetime of the MechanismRange instance.
+ */
+template <std::size_t NumFloatingPointFields, std::size_t NumDatumFields>
+struct MechanismRange {
+    MechanismRange(neuron::model_sorted_token const& cache_token,
+                   NrnThread&,
+                   Memb_list& ml,
+                   int type)
+        : MechanismRange{type, ml.get_storage_offset()} {
+        auto const& ptr_cache = mechanism::_get::_pdata_ptr_cache_data(cache_token, type);
+        m_pdata_ptrs = ptr_cache.data();
+        assert(ptr_cache.size() <= NumDatumFields);
+    }
+
+  protected:
+    MechanismRange(int mech_type, std::size_t offset)
```

Comment:
A (maybe stupid) question here.. Is it needed to have another `offset` if we already have an `int` number `mech_type` to use as an "offset"?

Reply:
I'm not sure I follow...we don't store the mechanism type (`mech_type`) in the struct (i.e. there is no `m_mech_type` member) 🤔

Reply:
Edit: initially `offset` looked trivial to me but I see it's more complicated than that.. I just haven't understood yet what this means for the mechanism data..

Reply:
Note that this is an internal/implementation detail constructor; it might make sense to start from the public constructors of MechanismRange and MechanismInstance, which delegate to it.

`offset` is an offset into the global vectors holding the mechanism data. If you have two threads then those are partitioned by thread; so the values for thread 0 start at offset 0 and the values for thread 1 start at some other offset. 
- In case of MechanismRange (roughly Memb_list; a range of mechanism instances), that's what this offset is storing. Then each thread will pass indices from 0..N-1 to the other accessors.
- In the case of MechanismInstance (wrapping access to a single mechanism instance), the offset is the index of that instance and we always pass 0 as the index to the other accessors.